### PR TITLE
Add early stopping and reporting based on validation data

### PR DESCRIPTION
**Pull request by dansbecker**
This PR allows a user to specify a validation dataset that are used for early stopping (and reporting). The PR was created to address [issue 85](https://github.com/tensorflow/skflow/issues/85)

I made changes in 3 places. 
1. The trainer now takes a dictionary containing the validation data (in the same format as the output of the data feeder's get_dict_fn). 
2. The fit method now takes arguments for val_X and val_y. It converts these into the correct format for the trainer.
3. The example file digits.py now uses early stopping, by supplying val_X and val_y.

I can add early stopping to other examples if this approach looks good, though their behavior should not otherwise be affected by the current PR.

### CODE REVIEW

**Comment by dansbecker** [comment]
The original PR fails the linter because `TensorFlowTrainer.train` has too many branches (14). I think this fairly represents a fair amount of complexity that's creeping into the train function.

Some of the reporting logic is currently in the `train` and some is in `_print_report`. 

I'd like to refactor reporting so all of the logic and printing are in a `TensorFlowTrain.report` method. I think this is a worthwhile change aside from the current `linter` error. But given that this change will resolve that error, do you mind I include it in the current PR?

**Comment by terrytangyuan** [comment]
@dansbecker This PR looks good to me. Regarding the lint error, you can disable locally by adding a comment like this `# pylint: disable=too-many-branches` to resolve it for now but we may need to refactor at some point. @ilblackdragon What do you think? Could you check this PR?

**Comment by dansbecker** [comment]
@terrytangyuan I updated the PR with the line you sent.  If you are open to the refactor (moving the output & reporting into a separate method), and I'll submit it as a separate PR.

**Comment by ilblackdragon** [comment]
Refactoring sounds good for the trainer.

I actually think that reporting may be even worth moving out of the trainer at all and making similar to "Monitor" interface (e.g. something you attach to the progress of training and can be customized) you mentioned in another thread. Then we can have a simple ConsoleMonitor that would produce current behaviour, and also, would allow extending with storing info to a local database or any other reporting tooling we want to add over time.

On the current PR: 
I'm a bit negative about adding more parameters to `fit()` method because this is not supported by all the other sci-kit learn tooling (like pipelines, cross validation etc).

Now scikit learn does have some usage for monitors, for [an example](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn.ensemble.GradientBoostingClassifier.fit) and they use it for early stopping as well.

Maybe we can have a ValidationMonitor(X_val, y_val) that would return true if model converged.

@dansbecker What do you think?

**Comment by dansbecker** [comment]
@ilblackdragon Now that you bring it up, adding arguments to the `fit()` method is obviously the wrong choice, for the reason you mention.

In the related [issue](https://github.com/tensorflow/skflow/issues/85), I also considered the option to create the monitor you mentioned, or to add an argument to the constructor of the `TensorFlowEstimator`that specifies what fraction of data is set aside for early stopping.

The monitor seems clearly more modular. Adding a frac_used_for_early_stopping at model specification seems to me to be an easier user interface.

If you confirm that you prefer the monitor, I'll do that.

**Comment by ilblackdragon** [comment]
Yeah, I saw your suggestion about a monitor and after thinking and looking at scikit learn I think it's a better way to go. We probably can also remove a lot of options from estimators if we move everything to monitors (like verbosity, savers and etc) - it's quite a few already and adding more won't scale.

So yes, on the monitors! :) Maybe let's make a `monitors` folder, because I think they will become a core part of APIs and there can be a lot of them.

Thanks for doing this, really appreciate your time and looking into options / trying things out!

**Comment by dansbecker** [comment]
Just committed the first version of the refactor.  This *appears* to work with early stopping on validation data.  I still need to 
[ ] Validate these results against the previous version and what's in master
[ ] Test the BaseMonitor that doesn't use validation data
[ ] Fix the printed output to show validation loss
[ ] Refactor again. There is currently a lot of duplicated code in both monitoring classes. I will abstract that code out.
[ ] Fix any code in examples that this breaks

So, it's definitely not ready yet... but there's an update showing the current status.

**Comment by terrytangyuan** [comment]
Looks great! BTW, you can reproduce the Travis error by `nosetests` for unit tests and `pylint skflow` for lint check.

**Comment by dansbecker** [comment]
@ilblackdragon @terrytangyuan This is ready for CR.

File: skflow/estimators/base.py [hunk 1]
```diff
@@ -197,6 +198,12 @@ def fit(self, X, y, logdir=None):
         self._data_feeder = setup_train_data_feeder(X, y,
                                                     self.n_classes,
                                                     self.batch_size)
+
+        if monitor is None:
+            self._monitor = monitors.BaseMonitor()
```
Comment (ilblackdragon) [inline skflow/estimators/base.py]: just a though may be for subsequent PR is to have `montiors.default_monitor()` function that returns this or another class. For example, this would allow to be able override default with `set_default_monitor()` or easily change which one is default if this line will be more then in one place.

File: skflow/monitors.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+class ValidationMonitor(BaseMonitor):
+    """Monitor that reports score for validation data and uses validation data for early stopping
+
+        val_X: Validation features
+        val_y: Validation labels
+        n_classes: Number of labels in output. 0 for regression
+        See BaseMonitor for arguments
+    """
+    def __init__(self, val_X, val_y, n_classes=0, *args, **kwargs):
+        super(ValidationMonitor, self).__init__()
```
Comment (ilblackdragon) [inline skflow/monitors.py]: what about BaseMonitor args, like verbosity, if one wants to change them?

File: skflow/monitors.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-few-public-methods
+# pylint: disable=too-many-arguments
+# pylint: disable=unused-argument
+# pylint: disable=attribute-defined-outside-init
+
+class BaseMonitor(object):
+    """ Base class for all learning monitors. Stores and reports training loss throughout learning
+
+        print_steps: Number of steps in between printing cost.
```
Comment (ilblackdragon) [inline skflow/monitors.py]: Covention for class comments:
```
Attributes:
   print_steps: ...
```

File: skflow/monitors.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    def create_val_feed_dict(self, inp, out):
+        """Set tensorflow placeholders and create validation data feed"""
+        self.val_dict = self.val_feeder.get_feed_dict_fn(inp, out)()
+
+    def _set_last_loss_seen(self):
+        """Sets self.last_loss_seen to most recent validation loss
+
+        Also stores this value to appropriate buffers
+        """
+        [val_loss] = self.sess.run([self.loss_expression_tensor], feed_dict=self.val_dict)
```
Comment (ilblackdragon) [inline skflow/monitors.py]: what if instead of passing inp/out/loss - pass the estimator itself, then it's possible to run predict and compute loss or any other function (for example if user wants to stop on accuracy) out of it.

not sure, what do you think about something like this?
```
def __init__(self, ..., n_classes=0, loss_fn=None)
    self.loss_fn = loss_fn
    ...

def update(self, estimator, ...):
     if self.loss_fn:
         self.loss_fn(estimator.predict(val_X), val_y)
    else:
         estimator.predict_loss(val_X, val_y) # this handles feed_dict and etc by a bit refactoring predict_ to support this case
```

where `loss_fn` if not None can be `sklearn.metrics.*`

I'm just a bit conscious that monitor has an interface that is very integrated with a concept of validation.

**Comment by ilblackdragon** [comment]
@dansbecker In general - looks pretty good, I like how trainer become a lot cleaner.
The only thing is to make monitor less validation centric. (e.g. one can make a monitor that would be maybe Ipython console or reporting into database, etc)

File: skflow/estimators/base.py (hunk 1 above)
Comment (dansbecker) [inline skflow/estimators/base.py]: Will include that in next commit.

File: skflow/monitors.py (hunk 2 above)
Comment (dansbecker) [inline skflow/monitors.py]: Thanks. Missing that was a bug. I meant to pass those args to the super class. Will include in next commit.

File: skflow/monitors.py (hunk 4 above)
Comment (dansbecker) [inline skflow/monitors.py]: The update is called from train, and the trainer doesn't currently keep track of that estimator.  So, I think passing the estimator may add more complexity than it would at first seem.

**Comment by dansbecker** [comment]
I agree that this ended up validation centric (in part because I originally intended to create this PR just to implement validation.) Do you think it's reasonable to address that as part of a subsequent PR, when we start adding more classical monitoring-type functionality?

File: skflow/monitors.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-few-public-methods
+# pylint: disable=too-many-arguments
+# pylint: disable=unused-argument
+# pylint: disable=attribute-defined-outside-init
+
+def default_monitor():
+    return(BaseMonitor())
```
Comment (ilblackdragon) [inline skflow/monitors.py]: nit: `return BaseMonitor()`

**Comment by ilblackdragon** [comment]
For this one, fix the linters and I think it's ok to merge it in and keep iterating (I don't think people will jump on using this API of monitors right away).

**Comment by dansbecker** [comment]
Fixed the lint. Let me know whether I should be squashing that type of commit. 

I think this PR closes #85
